odl/solvers/advanced/chambolle_pock.py

```diff
@@ -0,0 +1,278 @@
+# Copyright 2014-2016 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""First-order primal-dual algorithm developed by Chambolle and Pock."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+from builtins import super
+
+# External
+
+# Internal
+from odl.operator.operator import Operator
+from odl.operator.pspace_ops import ProductSpaceOperator
+
+__all__ = ('chambolle_pock_solver', 'f_cc_prox_l2_tv', 'g_prox_none')
+
+
+# TODO: add dual gap as convergence measure
+# TODO: diagonal preconditioning
+# TODO: add positivity constraint
+
+def chambolle_pock_solver(K, f_cc_prox, g_prox, sigma, tau, theta=1,
+                          niter=100, partial=None):
+    """Chambolle-Pock aglgorithms for convex optimization problems.
+
+    First order primal-dual hybrid-gradient (PDHG) method for non-smooth
+    convex optimization probelms with known saddle-point structure developed
+    by `Chambolle and Pock`_ (CP).
+
+
+    Parameters
+    ----------
+    K : `Operator` or `ProductSpaceOperator`
```

Comment:
> I don't like `K`. It was good to follow the lines of the paper while implementing, but there should be a better name for it. Don't use capital letters for variables. Pylint will complain, and rightly so.

You certainly have a point, but naming it is hard. The original paper does not provide any name except `K`. Perhaps this has some "standard" name in the literature, @aringh?

> Either we let the user provide `forward_op` and `sparsifying_op` (or shorter versions of those names) and combine them internally as I wrote earlier, or we need some good idea for a name.

Regarding the `forward_op`, `sparsifying_op` discussion, we would (imo) need to create another wrapper layer for these. A big unifying op is the only way to keep it general.

> Probably not. The question is is it obvious to the user that `Operator` is a special case of `ProductSpaceOperator`?

I think so, the name kinda implies it, no?